lgc/state/LgcContext.cpp

```diff
@@ -69,6 +70,21 @@ static cl::opt<bool> EmitLlvmBc("emit-llvm-bc", cl::desc("Emit LLVM bitcode inst
 static cl::opt<bool> EmitLgc("emit-lgc", cl::desc("Emit LLVM assembly suitable for input to LGC (middle-end compiler)"),
                              cl::init(false));
 
+// =====================================================================================================================
+// Set default for a command-line option, but only if command-line processing has not happened yet, or did not see
+// an occurrence of this option.
+//
+// @param name : Option name
+// @param value : Default option value
+template <typename T> static void setOptionDefault(const char *name, T value) {
+  auto optIterator = cl::getRegisteredOptions().find(name);
+  assert(optIterator != cl::getRegisteredOptions().end());
+  cl::Option *opt = optIterator->second;
+  if (opt->getNumOccurrences())
+    return;
+  *static_cast<cl::opt<T> *>(opt) = value;
```

Comment:
What happens when a wrong type is passed here? For example, the original option is `cl::opt<bool>` but `0` (int) is passes instead.

Reply:
It could crash horribly. Well, int vs bool would be ok, but e.g. int vs std::string would not.

Reply:
Is there a safer alternative then? Could we remove static from options' definitions and a list of `extern` options, or keep the statics but provide external storage for these options end make them extern instead?

Reply:
They're options in LLVM, and we don't really have a good justification for changing them upstream in that way.

I think it's much more likely that an upstream option would just disappear, rather than suddenly change type. In that case, this code silently does nothing.

Maybe I could make it safer by checking that the option is of the kind I expect. I could see whether that is possible.

Reply:
> Maybe I could make it safer by checking that the option is of the kind I expect. I could see whether that is possible.

This sounds like a good idea to me.
Another way would be to expose these configuration knobs through llvm pass creation function arguments.

With the cast as-is, this seems to me like a disaster waiting to happen.

Reply:
Found a typer-safer way of doing it.